src/mscorlib/corefx/System/Buffers/MultiProducerMultiConsumerBoundedQueue.cs

```diff
@@ -0,0 +1,172 @@
+                        T item = _slots[slotsIndex].Item;
+                        _slots[slotsIndex].Item = null;
+                        Volatile.Write(ref _slots[slotsIndex].SequenceNumber, currentHead + _slots.Length);
+                        return item;
+                    }
+                }
+                else if (delta < 0)
+                {
+                    // Queue is empty
+                    return null;
+                }
+
+                // Lost a race. Spin a bit, then try again.
+                spinner.SpinOnce();
+            }
+        }
+
+        /// <summary>
+        /// Attempts to enqueue the item.  If successful, the item will be stored
+        /// in the queue and true will be returned; otherwise, the item won't be stored, and false
+        /// will be returned.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool TryEnqueue(T item)
+        {
+            Debug.Assert(item != null);
+
+            // Loop in case of contention...
+            var spinner = new SpinWait();
+            while (true)
+            {
+                // Get the tail at which to try to return.
+                long currentTail = Volatile.Read(ref _headAndTail.Tail);
+                int slotsIndex = (int)(currentTail & _slotsMask);
+
+                // Read the sequence number for the tail position.
+                long sequenceNumber = Volatile.Read(ref _slots[slotsIndex].SequenceNumber);
+
+                // Sequence numbers are for returns, and are one higher for rents.
+                // So compare the sequence number to tail.  If they're the same, then this is our
+                // slot to return to, but we need to use an interlocked to advance the tail, as
+                // concurrent returners could take it out from under us. If the sequence number is less
+                // than tail, then there's no available slot, so we just drop the array.  If the sequence
+                // number is greater than tail, we lost a race with some other thread that incremented
+                // the sequence number, so we need to try again.
+                long delta = sequenceNumber - currentTail;
+                if (delta == 0)
+                {
+                    if (Interlocked.CompareExchange(ref _headAndTail.Tail, currentTail + 1, currentTail) == currentTail)
+                    {
+                        // Successfully reserved the slot.  Note that after the above CompareExchange, other threads
+                        // trying to return will end up spinning until we do the subsequent Write.
+                        _slots[slotsIndex].Item = item;
+                        Volatile.Write(ref _slots[slotsIndex].SequenceNumber, currentTail + 1);
+                        return true;
+                    }
+                }
+                else if (delta < 0)
+                {
+                    // Queue is full.
```

Comment:
But as in TryEnqueue, if old or torn values are read, this may also incorrectly determine that the queue is full and is not guaranteed to enqueue an item when there is space.